File: hot100/46. 全排列.py
# https://leetcode-cn.com/problems/permutations/
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ret = []

        def dfs(idx, path):
            if len(path) == n:
                ret.append(path)
                return
            for i in range(n):
                if nums[i] not in path:
                    dfs(i, path+[nums[i]])
        for i in range(n):
            dfs(i, [nums[i]])
        return ret


x = Solution()
print(x.permute([1, 2, 3]))

File: labuladong/46. 全排列.js
/** https://leetcode-cn.com/problems/permutations/
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
    const ret = [];
    function backtrack(cur = 0, path = []) {
        // 满足条件
        if (path.length === nums.length) {
            ret.push(path);
            return;
        }
        // 做出选择
        for (const n of nums) {
            if (!path.includes(n)) {
                backtrack(n, path.concat([n]));
            }
        }
    }

    for (const n of nums) {
        backtrack(n, [n]);
    }
    return ret;
};
